### PR TITLE
bump.sh checks if the images exist before updating config files

### PR DESCRIPTION
**Pull request by chizhg**
This is a fix to the issue 156032855

It uses the approach suggested in https://stackoverflow.com/a/55344819/13578870 to query the manifest url, to determine if the image with the given tag exists or not, which IMO is better than listing all image tags or using `docker pull`.

/assign @cjwagner @fejta

### CODE REVIEW

File: prow/cmd/autobump/bump.sh [hunk 1]
```diff
@@ -66,34 +66,52 @@ main() {
...
   bumpfiles+=("${CONFIG_PATH}")
   if [[ -n "${JOB_CONFIG_PATH}" ]]; then
     bumpfiles+=($(grep -rl -e "gcr.io/k8s-prow/" "${JOB_CONFIG_PATH}"))
   fi
 
-  echo -e "Attempting to bump the following files: ${bumpfiles[@]}" >&2
+  echo -e "Attempting to bump the following files:" >&2
+  IFS=$'\n'; echo -e "${bumpfiles[*]}" >&2
 
+  local token="$(gcloud config config-helper --format 'value(credential.access_token)')"
```
Comment (fejta) [inline prow/cmd/autobump/bump.sh:80]: `gcloud auth print-access-token`

File: prow/cmd/autobump/bump.sh [hunk 2]
```diff
@@ -66,34 +66,52 @@ main() {
...
+    local images="$(grep -o "${matcher}" "${file}")"
+    local array=(${images//\\n/})
+    for i in "${!array[@]}"; do
+      # image is in the format of gcr.io/k8s-prow/[image_name]:[tag]
+      local image="${array[i]}"
+      # Use the Docker Registry v2 API to query the image manifest to check if the given image tag exists or not.
+      # The manifest_url is in the format of https://gcr.io/v2/k8s-prow/[image_name]/manifests/[tag]
+      # Check more details from https://stackoverflow.com/a/55344819/13578870
+      local manifest_url=$(echo "$image" | sed "s/:/\/manifests\//" | sed "s/gcr.io/https:\/\/gcr.io\/v2/")
+      local code=$(curl -H "Authorization: Bearer $token" -o /dev/null -s -w "%{http_code}" "${manifest_url}")
```
Comment (fejta) [inline prow/cmd/autobump/bump.sh:94]: `--fail`? (may also want -L)

File: prow/cmd/autobump/bump.sh [hunk 3]
```diff
@@ -66,34 +66,52 @@ main() {
...
+  echo -e "Bumping: 'gcr.io/k8s-prow/' images to $(color-version "${new_version}") ..." >&2
 
   bumpfiles=($(add_suffix "$(split_on_commas "$COMPONENT_FILE_DIR")"))
   bumpfiles+=("${CONFIG_PATH}")
   if [[ -n "${JOB_CONFIG_PATH}" ]]; then
     bumpfiles+=($(grep -rl -e "gcr.io/k8s-prow/" "${JOB_CONFIG_PATH}"))
   fi
 
-  echo -e "Attempting to bump the following files: ${bumpfiles[@]}" >&2
+  echo -e "Attempting to bump the following files:" >&2
```
Comment (fejta) [inline prow/cmd/autobump/bump.sh:77]: Not using any escapes

File: prow/cmd/autobump/bump.sh [hunk 4]
```diff
@@ -66,34 +66,52 @@ main() {
...
+  IFS=$'\n'; echo -e "${bumpfiles[*]}" >&2
```
Comment (fejta) [inline prow/cmd/autobump/bump.sh:78]: wouldn't `IFS=$'\n' echo ...` be more appropriate than permanently changing IFS (via the `;`')? also

```sh
for bf in "${bumpfiles[@]}"; do
  echo -e "$bf"
done
```

is IMO easier to understand than working with IFS

File: prow/cmd/autobump/bump.sh [hunk 5]
```diff
@@ -66,34 +66,52 @@ main() {
...
+  local token="$(gcloud config config-helper --format 'value(credential.access_token)')"
   # Update image tags in the identified files.
-  filter="s/gcr.io\/k8s-prow\/\([[:alnum:]_-]\+\):v[a-f0-9-]\+/gcr.io\/k8s-prow\/\1:${new_version}/I"
+  local matcher="gcr.io\/k8s-prow\/\([[:alnum:]_-]\+\):v[a-f0-9-]\+"
+  local replacer="s/${matcher}/gcr.io\/k8s-prow\/\1:${new_version}/I"
   for file in "${bumpfiles[@]}"; do
-    ${SED} -i "${filter}" ${file}
+    local images="$(grep -o "${matcher}" "${file}")"
+    local array=(${images//\\n/})
+    for i in "${!array[@]}"; do
```
Comment (fejta) [inline prow/cmd/autobump/bump.sh:87]: No need for the index you're not doing anything with: `for image in "${array[@]}"; do`

File: prow/cmd/autobump/bump.sh (hunk 1 above)
Comment (chizhg) [reply prow/cmd/autobump/bump.sh]: Changed, thanks!

File: prow/cmd/autobump/bump.sh (hunk 2 above)
Comment (chizhg) [reply prow/cmd/autobump/bump.sh]: Added `--fail` and `-L`

File: prow/cmd/autobump/bump.sh (hunk 4 above)
Comment (chizhg) [reply prow/cmd/autobump/bump.sh]: Changed to for loop. Thanks!

File: prow/cmd/autobump/bump.sh (hunk 3 above)
Comment (chizhg) [reply prow/cmd/autobump/bump.sh]: Removed `-e` for this echo command.

File: prow/cmd/autobump/bump.sh (hunk 5 above)
Comment (chizhg) [reply prow/cmd/autobump/bump.sh:88]: I will get `array[*]: unbound variable` after changing to `for image in "${array[@]}"` when there is no matched image in the file. Not sure how to deal with it so I just kept as it is.

**Comment by chizhg** [comment]
> Not wild about investing much into this bash script
> 
> I'd prefer we work on making https://github.com/kubernetes/test-infra/tree/master/experiment/autobumper general purpose
> 
> https://github.com/kubernetes/test-infra/tree/master/experiment/image-bumper

Totally agree with that.
This will mainly be a short term prevention to the issue we had recently. In the long term we should do it with Go. I will look into how to unify these autobump tools later.

File: prow/cmd/autobump/bump.sh (hunk 5 above)
Comment (fejta) [reply prow/cmd/autobump/bump.sh:88]: This works just fine for me:


```sh
 set -o errexit
 set -o nounset
 set -o pipefail
 shopt -s dotglob
 shopt -s nullglob
 
 array=()
 for image in "${array[@]}"; do
     echo hi $image
 done
```

File: prow/cmd/autobump/bump.sh [hunk 6]
```diff
@@ -66,34 +66,54 @@ main() {
...
+      # image is in the format of gcr.io/k8s-prow/[image_name]:[tag]
+      local image="${array[i]}"
+      echo "Checking the existence of ${image} "
+      # Use the Docker Registry v2 API to query the image manifest to check if the given image tag exists or not.
+      # The manifest_url is in the format of https://gcr.io/v2/k8s-prow/[image_name]/manifests/[tag]
+      # Check more details from https://stackoverflow.com/a/55344819/13578870
+      local manifest_url=$(echo "$image" | sed "s/:/\/manifests\//" | sed "s/gcr.io/https:\/\/gcr.io\/v2/")
+      local code=$(curl --fail -L -H "Authorization: Bearer $token" -o /dev/null -s -w "%{http_code}" "${manifest_url}")
```
Comment (fejta) [inline prow/cmd/autobump/bump.sh:96]: IMO there's no need for the `-w %{http_code}` the `--fail` handles this.

File: prow/cmd/autobump/bump.sh (hunk 5 above)
Comment (chizhg) [reply prow/cmd/autobump/bump.sh]: It seems for different bash versions, the behavior is different for handling empty arrays.
I have changed to `for image in ${arr[@]+"${arr[@]}"};` as suggested in https://stackoverflow.com/questions/7577052/bash-empty-array-expansion-with-set-u, and it works.

File: prow/cmd/autobump/bump.sh (hunk 6 above)
Comment (chizhg) [reply prow/cmd/autobump/bump.sh]: Thanks, done.

**Comment by chizhg** [comment]
/hold cancel
It's now ready for a second review, thanks.

File: prow/cmd/autobump/bump.sh [hunk 7]
```diff
@@ -66,34 +66,52 @@ main() {
...
+    ${SED} -i "${replacer}" "${file}"
   done
 
   echo "bump.sh completed successfully!" >&2
 }
 
 check-args() {
   if [[ -z "${COMPONENT_FILE_DIR}" ]]; then
-    echo "ERROR: $COMPONENT_FILE_DIR must be specified." >&2
+    echo "ERROR: COMPONENT_FILE_DIR must be specified as an env var." >&2
```
Comment (fejta) [inline prow/cmd/autobump/bump.sh:108]: if you use ' instead of " then $COMPONENT_FILE_DIR will render correctly

**Pull request merged**